### PR TITLE
Update LevelDB to upstream commit f545dfabf

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#7911
- bitcoin/bitcoin#7982
- bitcoin/bitcoin#8133
- bitcoin/bitcoin#8784
  - Only the missing changes.
- bitcoin/bitcoin#8826
- bitcoin/bitcoin#8613
- bitcoin/bitcoin#10544
- bitcoin/bitcoin#10633
  - Only the changes to files and code we have.
- bitcoin/bitcoin#10806
- bitcoin/bitcoin#10958
- bitcoin/bitcoin#12451
- bitcoin/bitcoin#13925
- bitcoin/bitcoin#15270

This upgrades LevelDB in the exact same commit progression as upstream, up to January 2019.

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

**Comment by str4d** [comment]
There will be a second PR after this one to update LevelDB to match what Bitcoin Core currently uses, but that final migration in https://github.com/bitcoin/bitcoin/pull/17398 is quite large, so I figured I'd land these changes first.

**Comment by str4d** [comment]
Looks like there was a single transient failure in an otherwise successful CI run.

**Comment by daira** [comment]
Will review this today.

File: src/leveldb/db/db_bench.cc [hunk 1]
```diff
@@ -83,6 +84,14 @@ static bool FLAGS_histogram = false;
 // (initialized to default value by "main")
 static int FLAGS_write_buffer_size = 0;
 
+// Number of bytes written to each file.
+// (initialized to default value by "main")
+static int FLAGS_max_file_size = 0;
+
+// Approximate size of user data packed per block (before compression.
```
Comment (rex4539) [inline src/leveldb/db/db_bench.cc:91]:
```suggestion
// Approximate size of user data packed per block (before compression).
```

**Review by rex4539 — CHANGES_REQUESTED** [review]

File: src/leveldb/port/port_win.h [hunk 2]
```diff
@@ -32,9 +32,16 @@
...
 #define close _close
 #define fread_unlocked _fread_nolock
+#ifdef _WIN64
+#define ssize_t int64_t
+#else
+#define ssize_t int32_t
+#endif
 #endif
 
 #include <string>
```
Comment (daira) [inline src/leveldb/port/port_win.h:47]: Add below `#include <stdint.h>`:
```
static_assert(sizeof(ssize_t) == sizeof(size_t), "ssize_t should be the same size as size_t");
```

**Review by daira — APPROVED** [review]
ACK with minor comments.

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**